In [98]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
from sklearn import metrics
import warnings

# Ignore all warnings
warnings.filterwarnings("ignore")


In [41]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [44]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [45]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [46]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
data_scaled = StandardScaler().fit_transform(df_market_data)


In [47]:
# Create a DataFrame with the scaled data
df_scaled = pd.DataFrame(data_scaled, columns=df_market_data.columns)

# Copy the crypto names from the original data
df_scaled = df_scaled.set_index(df_market_data.index)

# Set the coinid column as index
df_scaled = df_scaled.rename_axis('coin_id')


# Display sample data
df_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

### Function to plot elbow curve

In [151]:
def plot_elbow_curve(df):
    # Create a list with the number of k-values from 1 to 11
    k = list(range(1,11))
    # Create an empty list to store the inertia values
    inertia = []


    # Create a for loop to compute the inertia with each possible value of k
    # Inside the loop:
    # 1. Create a KMeans model using the loop counter for the n_clusters
    # 2. Fit the model to the data using `df_market_data_scaled`
    # 3. Append the model.inertia_ to the inertia list
    for i in k:
        k_model = KMeans(n_clusters=i, random_state=1)
        k_model.fit(df)
        inertia.append(k_model.inertia_)
        
    # Create a dictionary with the data to plot the Elbow curve
    elbow_data = {"k": k, "inertia": inertia}


    # Create a DataFrame with the data to plot the Elbow curve
    df_elbow = pd.DataFrame(elbow_data)
    
    # Plot a line chart with all the inertia values computed with 
    # the different values of k to visually identify the optimal value for k.
    elbow_curve = df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
    )
    
    return elbow_curve

---

### Find the Best Value for k Using the Original Data.

In [152]:
# using the plot_elbow_curve function defined above to plot the elbow curve and find optimal k.

elbow_og = plot_elbow_curve(df_market_data)

elbow_og

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** The best value for k is 3.
            To determine the best number of clusters from my elbow curve, I need to look for the "elbow point," which is the point where the inertia (within-cluster sum of squares) starts to decrease at a slower rate. In this case, it seems the elbow point is likely to be around k=3. Beyond this point, the decrease in inertia is less steep, and the benefit of adding more clusters may not be significant.

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled Data

In [109]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=3, random_state=1)


In [110]:
# Fit the K-Means model using the scaled data
model.fit(df_scaled)

KMeans(n_clusters=3, random_state=1)

In [111]:
# Predict the clusters to group the cryptocurrencies using the scaled data
clusters = model.predict(df_scaled)

# Print the resulting array of cluster values.
print(clusters)

[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 0
 1 1 1 1]


In [112]:
# Create a copy of the DataFrame
df_market_predictions = df_market_data.copy()

In [113]:
# Add a new column to the DataFrame with the predicted clusters
df_market_predictions['cluster'] = clusters

# Display sample data
df_market_predictions.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                           1.08388                     7.60278   
ethereum                          0.22392                    10.38134   
tether                           -0.21173                     0.04935   
ripple                           -0.37819                    -0.60926   
bitcoin-cash                      2.90585                    17.09717   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                           6.57509                      7.67258   
ethereum                          4.80849                      0.13169   
tether                            0.00640                     -0.04237   
ripple                            2.24984                      0.23455   
bitcoin-cash                     14.75334                     15.74903   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                          -3.25185                      83.51840   
ethereum                        -12.88890                     186.77418   
tether                            0.28037                      -0.00542   
ripple                          -17.55245                      39.53888   
bitcoin-cash                    -13.71793                      21.66042   

              price_change_percentage_1y  cluster  
coin_id                                            
bitcoin                         37.51761        1  
ethereum                       101.96023        1  
tether                           0.01954        1  
ripple                         -16.60193        1  
bitcoin-cash                    14.49384        1

In [114]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
clusters_og = df_market_predictions.hvplot.scatter(
    x="price_change_percentage_24h", 
    y="price_change_percentage_7d", 
    by="cluster",
    hover_cols = 'coin_id'
)

clusters_og

:NdOverlay   [cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [115]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components = 3)

In [116]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
crypto_pca = pca.fit_transform(df_scaled)
# View the first five rows of the DataFrame. 
crypto_pca[:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [117]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

In [118]:
# finding the sum of the array
sum(pca.explained_variance_ratio_)

0.8950316570309841

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The 3 PCA components explain 89.5% of the total variance.

In [119]:
# Create a new DataFrame with the PCA data.
df_pca = pd.DataFrame(crypto_pca, columns = ['PCA1','PCA2','PCA3'])

# Copy the crypto names from the original data
df_pca = df_pca.set_index(df_market_data.index)

# Set the coinid column as index
df_pca = df_pca.rename_axis('coin_id')

# Display sample data
df_pca.head()

PCA1      PCA2      PCA3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [153]:
# using the plot_elbow_curve function defined above to plot the elbow curve and find optimal k.

elbow_pca = plot_elbow_curve(df_pca)

elbow_pca

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** The best value for k is 4.
            To determine the best number of clusters from my elbow curve, I need to look for the "elbow point," which is the point where the inertia (within-cluster sum of squares) starts to decrease at a slower rate. In this case, it seems the elbow point is likely to be around k=4. Beyond this point, the decrease in inertia is less steep, and the benefit of adding more clusters may not be significant.
            
            


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** Yes, the new elbow curve seems to flatten out around k = 4, where as with the original data, the curve seemed to flatten at k = 3. That being said, k = 3 would also be appropriate with the elbow curve above. The ultimate choice will also consider domain knowledge and other evaluation criteria.

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [143]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=1)


In [144]:
# Fit the K-Means model using the PCA data
model.fit(df_pca)


KMeans(n_clusters=4, random_state=1)

In [145]:
# Predict the clusters to group the cryptocurrencies using the PCA data
pca_clusters = model.predict(df_pca)

# Print the resulting array of cluster values.
pca_clusters

array([1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 1, 0, 0, 3, 0, 0, 0, 0])

In [146]:
# Create a copy of the DataFrame with the PCA data
df_pca_predictions = df_pca.copy()

# Add a new column to the DataFrame with the predicted clusters
df_pca_predictions['pca_cluster']=pca_clusters

# Display sample data
df_pca_predictions.head()

PCA1      PCA2      PCA3  pca_cluster
coin_id                                                
bitcoin      -0.600667  0.842760  0.461595            1
ethereum     -0.458261  0.458466  0.952877            1
tether       -0.433070 -0.168126 -0.641752            0
ripple       -0.471835 -0.222660 -0.479053            0
bitcoin-cash -1.157800  2.041209  1.859715            1

In [147]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
clusters_pca = df_pca_predictions.hvplot.scatter(
    x="PCA1", 
    y="PCA2", 
    by="pca_cluster",
    hover_cols = "coin_id"
)

clusters_pca

:NdOverlay   [pca_cluster]
   :Scatter   [PCA1]   (PCA2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [154]:
# Composite plot to contrast the Elbow curves
composite_elbow_plot = elbow_og + elbow_pca
composite_elbow_plot

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [149]:
# Composite plot to contrast the clusters
compostite_cluster_plot = clusters_og + clusters_pca
compostite_cluster_plot

:Layout
   .NdOverlay.I  :NdOverlay   [cluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [pca_cluster]
      :Scatter   [PCA1]   (PCA2,coin_id)

In [150]:
# Composite plot to contrast the clusters, with same x and y axis

df_test = df_market_data.copy()
df_test['pca_clusters'] = pca_clusters
clusters_test = df_test.hvplot.scatter(
    x="price_change_percentage_24h", 
    y="price_change_percentage_7d", 
    by="pca_clusters",
    hover_cols = 'coin_id'
)

compostite_cluster_plot = clusters_og + clusters_test
compostite_cluster_plot

:Layout
   .NdOverlay.I  :NdOverlay   [cluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [pca_clusters]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?
 
  * **Answer:** 
  
  Just by looking ath the composite graphs, there's a few things that stand out.
  
  
  - We already saw that the optimal no. of clusters went from 3 to 4, when using the original scaled data vs. the pca data. But it is important to note here, that 3 or 4 would have been a good choice for the pca clustering. 
  - Additionaly, the actual inertia values are much smaller for the pca elbow curve compared to the original. A smaller inertia indicates that the data points within each cluster are closer to their cluster centroid. It suggests that the clusters are more compact and tightly packed. Smaller inertia values are generally desirable as they indicate better separation of clusters.
  - When we look at the clusters based on the original data, all the coins except celcius-degree-token and ethlend are placed in the same cluster. The pca clusters are similar in the way that it places celcius-degree-token and ethlend in 2 different clusters, as before, but the rest of the coins are split by 2. Just as a note, if we conducted the pca analysis with k = 3, the clustering results will be the same as the original clustering results. 
  
  There are both positive and negative impacts of reducing dimensionality for k-means clustering:
  
  Positive
  
  -  It can make clustering more computationally efficient and reduce the risk of overfitting.

  - Fewer features may lead to simpler and more interpretable clusters, which can be useful for understanding the underlying structure of your data.

  - The K-Means algorithm may converge more quickly when working with fewer features, making it more efficient. This doesn't really apply to the dataset I'm using, which only has 41 rows.
  
  Negative
  
  - Reducing the number of features can result in a loss of information. 
  
  - Reducing the feature space without a clear understanding of the data's characteristics may lead to suboptimal clustering. 
  
  -  Fewer features can make the clustering algorithm more sensitive to outliers since there are fewer dimensions for the outliers to be diluted.
  
  
  